lib/utils/compose_ts.ts

```diff
@@ -730,6 +798,46 @@ export async function tarDirectory(
 	}
 }
 
+function setConfig(
+	configPath: string,
+	configData: MultiBuild.ParsedBalenaYml,
+	buildArgs: ParsedBuildArguments,
+): string {
+	const config = configData as any;
```

Comment:
By the way @joshbwlng, I wanted to mention the following, just so you are aware, but there is no need to change course on this PR, unless you really wanted to. In a product call last week, one of the topics was _"what to do about the potential for confusion of a balena app having two `balena.yml` files with different purposes:"_

```
./balena.yml - contracts / Deploy With Balena
./.balena/balena.yml - build secrets and variables
```

Check the **Info** and the **Timeline** tabs of this JF brainstorm topic:
https://jel.ly.fish/90f5a591-db7e-4a1a-8f72-1a38f203e3f2

A proposal was to move the build-variables section from `./.balena/balena.yml` to `./balena.yml`, which already has an `applicationEnvironmentVariables` section. But we would want to maintain backwards compatibility, so I envision a pre-processing step whereby any `build-variables` section of `./.balena/balena.yml` (in `PemissiveVarList` format) is translated to a less permissive format† and moved to `./balena.yml`. The `--buildArgs` logic could then operate on the less permissive format from `./balena.yml`. But there'd still be the complexity of the translation to be implemented. Also, if went that route, we would probably have to address issue #2183 (in relation to `balena push <ip-addr>`) before this PR could be merged. :-/

† Personally I prefer the dictionary format over the array format because it avoids the need of splitting strings on `=` is multiple places, which means spreading the logic of what to do if there are multiple `=` characters to multiple places. Also, as found in this PR, the dictionary format makes it easier to override variables of the same name with spread syntax for objects.

Anyway - I hadn't mentioned the contracts matter before because I thought it would just complicate and delay this PR, and it would be better to address the contract issues in other PRs. And I think it's indeed the case, but I leave it up to you. At least you are now aware of it. :-)

Reply:
If possible, I'd rather have those issues handled in separate PRs. I see the connection, but it feels like a bit too broad of a scope for what I'm trying to do here. And I'm on board with going the dictionary route, but still doing some debugging on how to properly handle both formats. When setting a global build variable `FIRST_VAR=0` in `.balena/balena.yml`, the parsed object becomes the following (kept some debug logs in for context):
```
initial global value: [
    "FIRST_VAR=0"
]

_.isArray(metadata['build-variables'].global): true

// Add in `SECOND_VAR` from buildArg

metadata: {
    "build-variables": {
        "global": {
            "0": "FIRST_VAR=0",
            "SECOND_VAR": "1234"
        },
        ....
    }
}
```

Reply:
> I'd rather have those issues handled in separate PRs

That's fine, I agree it's better. 👍 

> the parsed object becomes the following

`"0": "FIRST_VAR=0"` is a bug that you are debugging, right? It looks like a bug. :-)

> What I'm thinking is first converting to a dictionary if its an array, and then append/overwrite using values set with `--buildArgs`. What do you think?

Sounds like a good idea! 👍  Then, it may not even be necessary to convert back to array format (before adding the entry to the tar stream), even if the original format was an array. This is on the assumption that `resin-multibuild` accepts either format (array or dictionary). We should test this assumption with `balena push` to cloud and to device (or read `resin-multibuild` source code and satisfy ourselves that it is the case).

Reply:
@pdcastro Pushed a commit that seems to be working as expected - converting the metadata build variable sets into dictionaries if arrays, then appending/overwriting with `--buildArg` values. Tested with Livepush, trying cloud push out next.

Reply:
Cloud push worked as expected:
- valid NPM_TOKEN passed with `--buildArg` => was able to access private packages
- invalid NPM_TOKEN passed with `--buildArg` => unable to access private packages